File: solutions/cpp/kindergarten-garden/1/kindergarten_garden.cpp
#include "kindergarten_garden.h"

namespace kindergarten_garden {
    auto identify_plant(char letter) -> Plants {
        switch (letter) {
            case 'C': return Plants::clover;
            case 'G': return Plants::grass;
            case 'V': return Plants::violets;
            default: return Plants::radishes;
        }
    }

    auto plants(std::string_view cups, const char* child) -> std::array<Plants, 4> {
        const auto start_of_second_row = cups.size() / 2 + 1;
        
        const auto first_row_index = static_cast<unsigned>(*child - 'A') * 2;
        const auto second_row_index = start_of_second_row + first_row_index;
        
        return {identify_plant(cups[first_row_index]), 
                identify_plant(cups[first_row_index + 1]), 
                identify_plant(cups[second_row_index]), 
                identify_plant(cups[second_row_index + 1])};
    }
}  // namespace kindergarten_garden

File: solutions/javascript/kindergarten-garden/1/kindergarten-garden.js
const DEFAULT_STUDENTS = [
  'Alice',
  'Bob',
  'Charlie',
  'David',
  'Eve',
  'Fred',
  'Ginny',
  'Harriet',
  'Ileana',
  'Joseph',
  'Kincaid',
  'Larry',
];

const PLANT_CODES = {
  G: 'grass',
  V: 'violets',
  R: 'radishes',
  C: 'clover',
};

export class Garden {
  #map;
  
  #recurSplit(iter, n) {
    if (!iter.length) return [];
    return [iter.slice(0, n), ...this.#recurSplit(iter.slice(n), n)];
  }
  
  constructor(diagram, students = DEFAULT_STUDENTS) {
    const plants = diagram.split(/\s/).map(row => [...row].map(plant => PLANT_CODES[plant]));
    const splitRows = plants.map(row => this.#recurSplit(row, 2));
    const sortedStudents = students.sort((a, b) => a.localeCompare(b));
    const mapStudentToPlants = sortedStudents.map((student, idx) => {
      return [student, [...(splitRows?.[0]?.[idx] ?? []), ...(splitRows?.[1]?.[idx] ?? [])]];
    });
    this.#map = new Map(mapStudentToPlants);
  }

  plants(student) {
    return this.#map.get(student);
  }
}
